spec/models/job_proxy_dispatcher_spec.rb

```diff
@@ -172,5 +175,96 @@ module JobProxyDispatcherSpec
         end
       end
     end
+
+    context "with container and vms jobs" do
+      before(:each) do
+        @jobs = (@vms + @repo_vms + @container_images).collect(&:scan)
+        @dispatcher = JobProxyDispatcher.new
+      end
+
+      describe "#pending_jobs" do
+        it "returns only vm jobs by default" do
+          jobs = @dispatcher.pending_jobs
+          expect(jobs.count).to eq(@vms.count + @repo_vms.count)
+          expect(jobs.all? { |x| x.target_class == 'VmOrTemplate' }).to eq(true)
+          expect(jobs.count > 0).to be_truthy # in case something unexpected goes wrong
+        end
+
+        it "returns only container images jobs when requested" do
+          jobs = @dispatcher.pending_jobs(ContainerImage)
+          expect(jobs.count).to eq(@container_images.count)
+          expect(jobs.all? { |x| x.target_class == 'ContainerImage' }).to eq(true)
+          expect(jobs.count > 0).to be_truthy # in case something unexpected goes wrong
+        end
+      end
+
+      describe "#pending_container_jobs" do
+        it "returns container jobs by provider" do
+          jobs_by_ems, = @dispatcher.pending_container_jobs
+          expect(jobs_by_ems.keys).to match_array(@container_providers.map(&:id))
+          expect(jobs_by_ems[@container_providers.first.id].count).to eq(1)
+          expect(jobs_by_ems[@container_providers.second.id].count).to eq(2)
```

Comment:
@moolitayer I believe `@container_providers` is the result of creating them in the database, so it's not guaranteed to be in the same order.  This may become a sporadic failure.  I think you need to ask the `@container_providers.first` how many images it has and compare this to `jobs_by_ems[@container_providers.first.id].count`